### ISSUE TITLE
Updating global state with useGlobal inside useEffect in subscription callback

### ISSUE THREAD

**Issue by DankessS** [CharlesStover/reactn#131]
I have such piece of code:

```
const StepsWatcher = () => {
    const [stepsCount, setStepsCount] = useGlobal('stepsCount');
    const [isPedometerAvailable, setIsPedometerAvailable] = useState(false);

    useEffect(() => {
        Pedometer.isAvailableAsync().then(
            result => {
                setIsPedometerAvailable(result);

                if (!result) {
                    Alert.alert("Step Counter not available", "Step Counter is required to run this app");
                }
            },
            error => {
                setIsPedometerAvailable("Could not get Step Counter: " + error);
                Alert.alert("Step Counter error", "Could not get Step Counter: " + error);
            }
        );
        this._subscription = Pedometer.watchStepCount(result => {
            const stepsTaken = stepsCount + result.steps;
            setStepsCount(stepsTaken); //no effect, even with use useDispatch
            updatePlayerSteps(stepsTaken); //request to backend
        });

        return () => {
            this._subscription && this._subscription.remove();
            this._subscription = null;
        }

    }, []);


    return null;

};
```

I use Pedometer from expo-sensors to step tracking and when I am trying to update global state property with `setStepsCount(stepsTaken)` inside `Pedometer.watchStepCount` callback there is no effect. I know that I can pass `[stepsCount]` as a useEffect second patameter, but I don't want do this, becouse then every update my subscription `this._subscription` will be still created and cleaned. I tried to use useDispatch to change global state, but once again with no effect. I know that in React is an option to update for example state in this way(I mean use useState, not useGlobal): `setStepsCount(prevSteps => prevSteps + result.steps)` and there is no need to pass stepsCount do useEffect as an input, that's working fine. But I tried the same way with useGlobal, then I got an error, which means that there is no such way to do this with useGlobal. Could you give me some tips how to handle this?

**Comment by CharlesStover** [comment]
If you do not want to recreate the `Pedometer` subscription each render, dispatch would be the best bet over `setStepsCount`, as `setStepsCount` will require you have it and `stepsCount` as dependencies (just the nature of hooks!).

I would note that you should remove the `this.` before your `_subscription` as you are using arrow functions, which have no `this` context. `this` probably is referring to `window` or `global` in this case, so two instances of this component mounting simultaneously will share the same `this` and same subscription as a result.

Try this:

```javascript
// In your setup files
addReducer('incrementStepsCount', (state, _dispatch, count) => ({ stepsCount: state.stepsCount + count }));

// ...

// In your StepsWatch component
const increment = useDispatch('incrementStepsCount');

// ...

// In your watchStepCount
increment(result.steps);
```

I typed this stream-of-conscious, so you may need to correct any mistakes, but otherwise let me know if this solves your issue.

**Comment by DankessS** [comment]
I just handled the problem in several dozens minutes before your repond. I realized that result.steps in Pedometer.watchStepCount returns sum of taken steps, so I don't need to sum steps on my own. I can just use setNewStepsCount(result.steps), this properly updates global state of stepsCount and now I don't need additionally use of stepsCount global value inside Pedometer.watchStepCount (which is in fact not a useEffect hook dependencies, cause of to often change). 

Thank you for your response, I will check it also, but I think this issue can be closed.
